File: leetcode-solutions/ruby/101. Symmetric Tree.rb
# https://leetcode.com/problems/symmetric-tree/

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_symmetric(root)
  return true if root.nil? || (root.left.nil? && root.right.nil?)
  
  is_symatric(root.left, root.right)
end

def is_symatric(node1, node2)
  return true if node1.nil? && node2.nil?
  # puts "1 #{node1.val}, #{node2.val}"
  if node1 && node2 && (node1.val == node2.val)
      return is_symatric(node1.left, node2.right) && is_symatric(node1.right, node2.left)    
  else
      return false
  end
end

File: leetcode-solutions/python/101. Symmetric Tree.py
# https://leetcode.com/problems/symmetric-tree/

from collections import deque

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None: return True
        def same(r1, r2):
            if r1 is None and r2 is None:
                return True
            if r1 is None or r2 is None:
                return False
            
            return r1.val == r2.val and same(r1.left, r2.right) and same(r1.right, r2.left)
        return same(root.left, root.right)
